src/common/sorting/hashed_sort.cpp

```diff
@@ -0,0 +1,482 @@
+                                                  const Orders &order_bys,
+                                                  const vector<unique_ptr<BaseStatistics>> &partition_stats) {
+
+	// we sort by both 1) partition by expression list and 2) order by expressions
+	const auto partition_cols = partition_bys.size();
+	for (idx_t prt_idx = 0; prt_idx < partition_cols; prt_idx++) {
+		auto &pexpr = partition_bys[prt_idx];
+
+		if (partition_stats.empty() || !partition_stats[prt_idx]) {
+			orders.emplace_back(OrderType::ASCENDING, OrderByNullType::NULLS_FIRST, pexpr->Copy(), nullptr);
+		} else {
+			orders.emplace_back(OrderType::ASCENDING, OrderByNullType::NULLS_FIRST, pexpr->Copy(),
+			                    partition_stats[prt_idx]->ToUnique());
+		}
+		partitions.emplace_back(orders.back().Copy());
+	}
+
+	for (const auto &order : order_bys) {
+		orders.emplace_back(order.Copy());
+	}
+}
+
+HashedSortGlobalSinkState::HashedSortGlobalSinkState(ClientContext &context,
+                                                     const vector<unique_ptr<Expression>> &partition_bys,
+                                                     const vector<BoundOrderByNode> &order_bys,
+                                                     const Types &input_types,
+                                                     const vector<unique_ptr<BaseStatistics>> &partition_stats,
+                                                     idx_t estimated_cardinality)
+    : context(context), buffer_manager(BufferManager::GetBufferManager(context)), allocator(Allocator::Get(context)),
+      fixed_bits(0), payload_types(input_types), max_bits(1), count(0) {
+
+	GenerateOrderings(partitions, orders, partition_bys, order_bys, partition_stats);
+
+	// The payload prefix is the same as the input schema
+	for (column_t i = 0; i < payload_types.size(); ++i) {
+		scan_ids.emplace_back(i);
+	}
+
+	//	We have to compute ordering expressions ourselves and materialise them.
+	//	To do this, we scan the orders and add generate extra payload columns that we can reference.
+	for (auto &order : orders) {
+		auto &expr = *order.expression;
+		if (expr.GetExpressionClass() == ExpressionClass::BOUND_REF) {
+			auto &ref = expr.Cast<BoundReferenceExpression>();
+			sort_ids.emplace_back(ref.index);
+			continue;
+		}
+
+		//	Real expression - replace with a ref and save the expression
+		auto saved = std::move(order.expression);
+		const auto type = saved->return_type;
+		const auto idx = payload_types.size();
+		order.expression = make_uniq<BoundReferenceExpression>(type, idx);
+		sort_ids.emplace_back(idx);
+		payload_types.emplace_back(type);
+		sort_exprs.emplace_back(std::move(saved));
+	}
+
+	const auto memory_per_thread = PhysicalOperator::GetMaxThreadMemory(context);
+	const auto thread_pages = PreviousPowerOfTwo(memory_per_thread / (4 * buffer_manager.GetBlockAllocSize()));
+	while (max_bits < 10 && (thread_pages >> max_bits) > 1) {
+		++max_bits;
```

Comment:
I would consider bounding `max_bits` to 8. I've experimented a lot with this in the grouped hash aggregate, and when partitioning to more than 256 partitions, the TLB cache misses will cause performance to degrade. 256 partitions is enough to handle most workloads

Reply:
Interesting! This is left over from the TUM paper, but that is 10 years old and I don't think they checked that.